File: backend/src/controllers/searchController.js
/**
 * searchController.js
 * Handles POST /search
 *
 * DESIGN DECISION — Why not write directly to DB here?
 * Because under high load (thousands of concurrent searches),
 * synchronous DB writes create a write bottleneck.
 * Instead, we push to the in-memory batch buffer (addToBuffer),
 * which aggregates and flushes in bulk.
 */

import { addToBuffer } from '../workers/batchWriter.js';

export async function search(req, res) {
  const { query } = req.body;

  if (!query || typeof query !== 'string' || !query.trim()) {
    return res.status(400).json({ error: 'query field is required' });
  }

  const normalizedQuery = query.trim().toLowerCase();

  // Add to in-memory buffer — non-blocking, returns immediately
  addToBuffer(normalizedQuery);

  res.json({ message: 'Searched', query: normalizedQuery });
}

File: backend-java/src/main/java/com/searchiq/controller/SearchController.java
package com.searchiq.controller;

import com.searchiq.worker.BatchWriterService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

/**
 * SearchController — handles POST /api/search
 *
 * DESIGN DECISION — Why not write directly to MongoDB here?
 * Under high load (thousands of concurrent searches), synchronous DB writes
 * create a write bottleneck. Instead, we push to the in-memory batch buffer.
 * The buffer aggregates and flushes to MongoDB in bulk every 30 seconds.
 *
 * This endpoint returns immediately — O(1) time, no DB call.
 */
@RestController
@RequestMapping("/api")
public class SearchController {

    private final BatchWriterService batchWriter;

    public SearchController(BatchWriterService batchWriter) {
        this.batchWriter = batchWriter;
    }

    /**
     * POST /api/search
     * Body: { "query": "iphone 15" }
     * Returns: { "message": "Searched", "query": "iphone 15" }
     */
    @PostMapping("/search")
    public ResponseEntity<Map<String, Object>> search(
            @RequestBody Map<String, String> body) {

        String query = body.get("query");

        if (query == null || query.isBlank()) {
            return ResponseEntity.badRequest()
                    .body(Map.of("error", "query field is required"));
        }

        String normalized = query.trim().toLowerCase();

        // Add to in-memory buffer — non-blocking, returns immediately
        batchWriter.addToBuffer(normalized);

        return ResponseEntity.ok(Map.of(
                "message", "Searched",
                "query",   normalized
        ));
    }
}
